projects/packages/mcp/src/AbilitiesRegistry/Executors/Post/PostGetExecutor.php

```diff
@@ -0,0 +1,254 @@
+				if ( ctype_digit( (string) $target_site ) ) {
+					$target_blog_id = (int) $target_site;
+
+					// Validate blog ID.
+					if ( $target_blog_id < 1 ) {
+						return array(
+							'success' => false,
+							'error'   => array(
+								'code'    => 'invalid_site_id',
+								'message' => 'Invalid site ID provided',
+								'status'  => 400,
+							),
+						);
+					}
+
+					$blog_details = get_blog_details( $target_blog_id );
+				} else {
+					// It's a URL - decode and get blog details.
+					$target_site = urldecode( $target_site );
+					$target_site = str_replace( '::', '/', $target_site );
+
+					$blog_details = apply_filters( 'jetpack_mcp_get_blog_details_for_url', null, $target_site );
+					if ( ! $blog_details ) {
+						return array(
+							'success' => false,
+							'error'   => array(
+								'code'    => 'site_not_found',
+								'message' => 'Site not found or inaccessible',
+								'status'  => 404,
+							),
+						);
+					}
+				}
+
+				// Validate blog details.
+				if ( ! $blog_details || is_wp_error( $blog_details ) ) {
+					return array(
+						'success' => false,
+						'error'   => array(
+							'code'    => 'site_not_found',
+							'message' => 'Site not found or inaccessible',
+							'status'  => 404,
+						),
+					);
+				}
+
+				$target_blog_id = (int) $blog_details->blog_id;
+
+				// Only switch if it's a different site.
+				if ( $target_blog_id !== $original_blog_id ) {
+					// Check if site is restricted.
+					$is_suspended = apply_filters( 'jetpack_mcp_is_site_suspended', false, $target_blog_id );
+					if ( $is_suspended ) {
+						return array(
+							'success' => false,
+							'error'   => array(
+								'code'    => 'site_suspended',
+								'message' => 'This site has been suspended',
+								'status'  => 403,
+							),
+						);
+					}
+
+					// Check if site is confidential.
+					$is_ai_access_allowed = apply_filters( 'jetpack_site_ai_access_allowed', false, $target_blog_id );
+					if ( ! $is_ai_access_allowed ) {
+						return array(
+							'success' => false,
+							'error'   => array(
+								'code'    => 'site_confidential',
+								'message' => 'This site is confidential and cannot be accessed',
+								'status'  => 403,
+							),
+						);
+					}
+
+					switch_to_blog( $target_blog_id );
+					$switched = true;
+
+					// Trigger action for any additional setup needed.
+					if ( has_action( 'rest_api_switched_to_blog' ) ) {
+						do_action( 'rest_api_switched_to_blog' );
+					}
+				}
+			}
+
+			// Get single post using helper.
+			$post_data = PostQueryHelper::get_single_post( $input );
+
+			if ( ! $post_data ) {
+				return array(
+					'success' => false,
+					'error'   => array(
+						'code'    => 'post_not_found',
+						'message' => 'Post not found or inaccessible',
+						'status'  => 404,
+					),
+				);
+			}
+
+			// Check AI access for the current site (when not switching)
+			$current_blog_id      = get_current_blog_id();
```

Comment:
The check happens after data has been retrieved, we could move it before any data access.